Added support for waiting promises returned by the abstract function

stochris opened the pull request:
I rewrote the logic, so it could wait on promises, then async call the next state, and so on.

Commit:
Added support for waiting promises returned by the abstract function

stochris commented:
One possible caveat might be that triggering a second transition before the first transition was resolved, due to closure, the top level variable might be rewritten

nonplus commented:
What _should_ the behavior be if a second transition is triggered while the first one is in progress?  I would expect the second to override the first one.

The following needs to be addressed:
- Don't check in changes to the top-level `angular-ui-router-default.js`.  This file is generated from `src/angular-ui-router-default.js` and checked in whenever I make a new release (`gulp build`).
- The `"with abstract = true"` unit tests no longer verify that exception is thrown.  If no exception is thrown by `$state.go()`, the first test will hang (`done()` not called) and the second one will incorrectly report success.  Also, please fix the typo I made in the test names, e.g. `"should fails"` -> `"should fail transition to base.abstract"` and `"should fails to ^"` -> `"should fail transition to ^"`
- We need unit tests similar to `"with abstract = function() { return ...; }"` for _resolved_ promises, e.g. `"with abstract = function() { return resolvedPromise; }"`
- We need unit tests similar to `"with abstract = function() { return ...; }"` for _rejected_ promises, e.g. `"with abstract = function() { return rejectedPromise; }"`

stochris commented:
I agree. As with ui-router behavior, the second transition should override the first.
Regarding the problems you noticed : 
- I'll undo the changes to the top level file
- I'll review the `"with abstract  = true"` test, and make sure it's fully working
- I just added a test, that checks a resolved promise
- I'll add more use cases ASAP

stochris commented:
- Now calling another transition will make the old promise be rejected with the same message ui-router uses ( `transition superseded`) 
- I added 2 more tests, checking when a promise is rejected, and when a transition is superseded. Should I add more ?
- I reverted the build file back to the original
- I fixed the `with abstract = true` tests. Now they work, and throw errors and intended.

nonplus commented:
One more thing and then it's good to go.
- The `'with reject promise from abstract'` unit test should assert that after the promise is rejected, the current `$state` has not changed.

nonplus commented:
Also, please squash your commits into one.

stochris commented:
I added a check for the current state in `with reject promise from abstract` unit test.

Thanks for all the feedback :smile:

stochris commented:
I was wondering about the possibility of adding more extensibility to the returned value of the abstract property / function.

I'm thinking something along the lines of [this](https://gist.github.com/stochris/ec90022f1519e1b7ccafc26ed46e7f86)
